src/Build/Graph/ProjectGraph.cs

```diff
@@ -75,6 +76,16 @@ public sealed class ProjectGraph
 
         public IReadOnlyCollection<ProjectGraphNode> GraphRoots { get; }
 
+        /// <summary>
+        ///     Gets a distinct enumeration of all of the environment variables reads during evaluation.
+        /// </summary>
+        public IEnumerable<string> EnvironmentVariableReads => _projectNodesTopologicallySorted?.Value.Select(p => p.ProjectInstance).SelectMany(pi => pi.EnvironmentVariableReads).Distinct();
```

Comment:
>EnvironmentVariableReads [](start = 154, length = 24)

This may be premature optimization, so feel free to ignore my comment, but it occurs to me that there is going to be a relatively big set of common env variables that are the same across every project (platform, configuration, etc).
So by having each project carry its own set, and aggregating that at the project graph level could increase memory footprint unnecessary. Each project could report reads an uninitialize vars to the project graph as they are being discovered, so each item will be there only once, and they can be added to a set or similar so de deduplication will also be unnecessary.

Reply:
I've changed this aggregation to perform better (not leaving it to LINQ operations) in more recent commits. Others have suggested using an event mechanism, but as I responded below, I'm not sure how to nicely plumb such a system into the current architecture where it would have pass through ProjectGraph --> GraphBuilder --> ProjectInstance.